libexec/rtld-elf/rtld_c18n.c

```diff
@@ -1612,268 +1575,504 @@ _rtld_thread_start_init(void (*p)(struct pthread *))
+	bool siginfo;
+	void *sigfunc;
+	struct tramp_header *header;
 	const Obj_Entry *defobj;
+	compart_id_t callee;
+	stk_table_index callee_idx;
+	struct stk_table *table;
+	struct sigframe *osp, *nsp;
+	struct trusted_frame *tf, *ntf;
+	sigset_t oset;
+
+	siginfo = (act->sa_flags & SA_SIGINFO) != 0;
+	if (siginfo)
+		sigfunc = act->sa_sigaction;
+	else
+		sigfunc = act->sa_handler;
+	if (!cheri_gettag(sigfunc)) {
+		rtld_fdprintf(STDERR_FILENO,
+		    "c18n: Invalid handler %#p for signal %d\n",
+		    sigfunc, sig);
+		abort();
+	}
 
 	/*
 	 * If the signal handler is not already wrapped by a trampoline, wrap it
 	 * in one.
 	 */
+	header = tramp_reflect(sigfunc);
 	if (header == NULL) {
-		defobj = obj_from_addr(context);
-
-		/*
-		 * If SA_SIGINFO is not set, the signal handler can have one of
-		 * two possible signatures.
-		 */
-		if ((act->sa_flags & SA_SIGINFO) == 0)
-			fsig = (struct func_sig) { .valid = false };
-		else
-			fsig = (struct func_sig) {
-				.valid = true,
-				.reg_args = 3, .mem_args = false,
-				.ret_args = NONE
-			};
-
-		context = tramp_intern(NULL, &(struct tramp_data) {
-		    .target = context,
-		    .defobj = defobj,
-		    .sig = fsig
+		defobj = obj_from_addr(sigfunc);
+		sigfunc = tramp_intern(NULL, &(struct tramp_data) {
+		    .target = sigfunc,
+		    .defobj = defobj
 		});
 	} else
 		defobj = header->defobj;
+	callee = defobj->compart_id;
+	callee_idx = cid_to_index(callee);
 
 	/*
-	 * XXX: Enforce signal handling policy. We need to determine who is
-	 * registering the signal, who is handling the signal, and ensure both
-	 * are allowed.
+	 * Get the stack of the signal handler's compartment.
 	 */
-	(void)sig;
-	if (defobj->compart_id == C18N_RTLD_COMPART_ID)
-		rtld_fatal("c18n: Attempting to register an RTLD function as "
-		    "a signal handler");
+	table = get_stk_table();
+	osp = get_or_create_untrusted_stk(callee, &table);
+	nsp = osp - 1;
+	*nsp = (struct sigframe) {
+		.sf_si = *info,
+		.sf_uc = *ucp
+	};
+	table->data[callee].stack = nsp;
 
-	act->sa_flags |= SA_SIGINFO;
 	/*
-	 * XXX: Ignore sigaltstack for now.
+	 * Push a dummy frame onto the trusted stack that would restore the
+	 * stack top of the signal handler's compartment during an unwind.
 	 */
-	act->sa_flags &= ~SA_ONSTACK;
+	tf = get_trusted_stk();
+	ntf = tf - 1;
+	*ntf = (struct trusted_frame) {
+		.osp = osp,
+		.previous = tf,
+		.caller = callee_idx,
+		/*
+		 * This field is used by the next trampoline to determine the
+		 * current compartment.
+		 */
+		.callee = cid_to_index(RTLD_COMPART_ID)
+	};
+	set_trusted_stk(ntf);
 
-	return (context);
-}
+	/*
+	 * Restore the user-supplied signal mask before invoking the handler.
+	 */
+	sigprocmask(SIG_SETMASK, &act->sa_mask, &oset);
 
-void _rtld_sigaction_end(int, void *, const struct sigaction *,
-    struct sigaction *);
+	if (siginfo)
+		((__siginfohandler_t *)sigfunc)(sig, &nsp->sf_si, &nsp->sf_uc);
+	else
+		((__sighandler_t *)sigfunc)(sig);
```

Comment:
Should we just use the "old" calling convention in the non-siginfo case?

Reply:
Calling `sigfunc` with the BSD-style signature will cause problems in the future where function pointers carry signature information so that the caller and the callee agree on the exact number of arguments being passed. Plus, `man sigaction` says that the BSD-style signature is deprecated. So I think it's probably not worth supporting anymore.

Reply:
In the future will we be able to query the number of arguments from the signature information to choose the right variant?  The kernel still always uses the BSD-style (since on current architectures it works for both ABIs) and presumably we will have to teach sigsend in the kernel to check the signature information.

Reply:
Why can't we just ignore the signature and always put the arguments there?

Reply:
> Why can't we just ignore the signature and always put the arguments there?

We can for now, but once we implement function pointer signature checks, this will fail to work.

Going back to @bsdjhb's question, yes it will be possible to dynamically query the signature of a function pointer, but here it is not really necessary because the the c18n wrapper of `sigaction` always sets the `SA_SIGINFO` flag, so the kernel will always invoke with the POSIX-style signature.

Reply:
Yes but my point was more that you could query the signature both here and in the kernel to DTRT in both places.  That is to say, I think in general you probably want to match the behavior of sendsig here to avoid surprises.

Reply:
Hmmm, it seems that arm64's `sendsig` has already dropped support for the old BSD-style signature?

Reply:
Interesting. I don't know if that was deliberate or an oversight.

Reply:
Probably deliberate.  RISC-V doesn't have it either.  I almost said earlier that the right place to have dropped support for it was when adding new architectures like aarch64 and RISC-V.  I feel like I had to patch this in CheriBSD at one point, but perhaps that was only for MIPS.